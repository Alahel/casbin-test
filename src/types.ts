import type { Enforcer } from "casbin"
import type { Knex } from "knex"

export type PolicySub = string

export type Enforce = {
	ef: Enforcer
	sub: PolicySub
	obj: PolicyObj
	act: PolicyAct
	eft?: PolicyEft
	log?: boolean
	explicit?: boolean
}

export type Context = {
	knex: Knex
	ef: Enforcer
}

export type PolicyTuple = [PolicySub, PolicyObj | "*", PolicyAct | "*", PolicyEft]
export const PolicyEftRetrieveStartIndex = 0
export const PolicyEftIndex = 3

// grouping of groups inside groups, or groups associated to roles, or users associated to roles
export type GroupingPolicyTuple = [PolicySub, PolicySub]

export enum PolicyObj {
	Movie = "movie",
	MovieLocalizedData = "movie_localized_data",
	MovieBrandedData = "movie_branded_data",
	Theater = "theater",
	ConfigurationImport = "configuration_import",
}

export enum PolicyAct {
	Create = "create",
	Read = "read",
	Update = "update",
	Delete = "delete",
	Export = "export",
	Execute = "execute",
}

export enum PolicyEft {
	Allow = "allow",
	Deny = "deny",
}

// ---shape of basic policies---
export const PolicyCRUD = [PolicyAct.Create, PolicyAct.Read, PolicyAct.Update, PolicyAct.Delete] as const

// todo: use those restricted permissions per objects
export const PolicyActPerObj = {
	[PolicyObj.Movie]: PolicyCRUD,
	[PolicyObj.MovieLocalizedData]: PolicyCRUD,
	[PolicyObj.MovieBrandedData]: PolicyCRUD,
	[PolicyObj.Theater]: [...PolicyCRUD, PolicyAct.Export],
	[PolicyObj.ConfigurationImport]: [PolicyAct.Execute],
} as const
