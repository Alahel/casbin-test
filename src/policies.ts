import { PolicyActionEnum, PolicyRessourceEnum } from "./types"

// ---shape of basic policies---
export const PolicyCRUD = [
	PolicyActionEnum.CREATE,
	PolicyActionEnum.READ,
	PolicyActionEnum.UPDATE,
	PolicyActionEnum.DELETE,
] as const

export const PolicyAvailablePermissions = {
	[PolicyRessourceEnum.MOVIE]: PolicyCRUD,
	[PolicyRessourceEnum.MOVIE_LOCALIZED_DATA]: PolicyCRUD,
	[PolicyRessourceEnum.MOVIE_BRANDED_DATA]: PolicyCRUD,
	[PolicyRessourceEnum.THEATER]: [PolicyActionEnum.READ],
} as const

// ---user policies---
export const POLICY_USER_ALICE = "PolicyUser:alice"
export const POLICY_USER_BOB = "PolicyUser:bob"
export const POLICY_USER_JOHN = "PolicyUser:john"

// ---roles policies---
export const POLICY_ROLE_REDAC = "PolicyRole:redac"
export const POLICY_ROLE_B2B_MAVENS = "PolicyRole:b2b_mavens"

// ---group policies---
export const POLICY_GROUP_REDAC_FR = "PolicyGroup:redac_fr"
export const POLICY_GROUP_B2B_RIDGEFIELD = "PolicyGroup:b2b_ridgefield"
export const POLICY_GROUP_B2B_APPDOT = "PolicyGroup:b2b_appdot"

export const ALL_POLICIES = [
	// roles
	[POLICY_ROLE_REDAC, PolicyRessourceEnum.MOVIE, PolicyActionEnum.READ],
	[POLICY_ROLE_REDAC, PolicyRessourceEnum.MOVIE_BRANDED_DATA, PolicyActionEnum.CREATE],
	[POLICY_ROLE_REDAC, PolicyRessourceEnum.MOVIE_BRANDED_DATA, PolicyActionEnum.READ],
	[POLICY_ROLE_REDAC, PolicyRessourceEnum.MOVIE_BRANDED_DATA, PolicyActionEnum.UPDATE],
	[POLICY_ROLE_REDAC, PolicyRessourceEnum.MOVIE_BRANDED_DATA, PolicyActionEnum.DELETE],

	[POLICY_ROLE_B2B_MAVENS, PolicyRessourceEnum.MOVIE, PolicyActionEnum.READ],
	[POLICY_ROLE_B2B_MAVENS, PolicyRessourceEnum.MOVIE_LOCALIZED_DATA, PolicyActionEnum.CREATE],
	[POLICY_ROLE_B2B_MAVENS, PolicyRessourceEnum.MOVIE_LOCALIZED_DATA, PolicyActionEnum.READ],
	[POLICY_ROLE_B2B_MAVENS, PolicyRessourceEnum.MOVIE_LOCALIZED_DATA, PolicyActionEnum.UPDATE],
	[POLICY_ROLE_B2B_MAVENS, PolicyRessourceEnum.MOVIE_LOCALIZED_DATA, PolicyActionEnum.DELETE],

	// user
	[POLICY_USER_ALICE, PolicyRessourceEnum.MOVIE, PolicyActionEnum.DELETE],

	// group
	[POLICY_GROUP_B2B_RIDGEFIELD, PolicyRessourceEnum.MOVIE, PolicyActionEnum.UPDATE],
]

export const ALL_GROUPING_POLICIES = [
	// roles of groups
	[POLICY_GROUP_REDAC_FR, POLICY_ROLE_REDAC],
	[POLICY_GROUP_B2B_RIDGEFIELD, POLICY_ROLE_B2B_MAVENS],

	// users in groups
	[POLICY_USER_ALICE, POLICY_GROUP_REDAC_FR],
	[POLICY_USER_BOB, POLICY_GROUP_B2B_RIDGEFIELD],
	[POLICY_USER_JOHN, POLICY_GROUP_B2B_APPDOT],

	// groups inheritance
	[POLICY_GROUP_B2B_APPDOT, POLICY_GROUP_REDAC_FR],
]
