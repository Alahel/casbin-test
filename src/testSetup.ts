import type { Enforcer } from "casbin"
import type { Knex } from "knex"
import { afterAll, beforeAll, expect, it } from "vitest"
import { mainFromDB } from "./bootstrap.js"
import { policyEnforceCheck } from "./enforce.js"
import { type Enforce, type PolicyAct, PolicyEft } from "./types.js"

let ef: Enforcer
let knex: Knex
let policyCheck: ReturnType<typeof policyEnforceCheck>
beforeAll(async () => {
	const init = await mainFromDB()
	ef = init.ef
	knex = init.knex
	policyCheck = policyEnforceCheck({ ef })
})

afterAll(async () => {
	await knex.destroy()
})

export type TestCtx = {
	ef: Enforcer
	knex: Knex
	policyCheck: ReturnType<typeof policyEnforceCheck>
}
export const getTestCtx = async (): Promise<TestCtx> => ({ ef, knex, policyCheck })

export const assertPermission =
	({ sub, obj, act, eft = PolicyEft.Allow, explicit }: Omit<Enforce, "log" | "ef">) =>
	(expectedResult = true) => {
		it(`check permission for "${sub}" to ${eft === PolicyEft.Allow ? "perform" : "prevent"} "${act}" on "${obj}", ${explicit ? "explicitely" : "with inheritance"}, expected to be "${expectedResult}"`, async () => {
			expect(
				await policyCheck({
					sub,
					obj,
					act,
					eft,
				}),
			).toBe(expectedResult)
		})
	}

export const assertPermissions =
	({ sub, obj, eft = PolicyEft.Allow, explicit }: Omit<Enforce, "log" | "ef" | "act">) =>
	({ acts }: { acts: readonly PolicyAct[] }) =>
	(expectedResult = true) => {
		for (const act of acts) assertPermission({ sub, obj, act, eft, explicit })(expectedResult)
	}
