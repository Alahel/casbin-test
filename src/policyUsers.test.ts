import type { Enforcer } from "casbin"
import { beforeAll, describe, expect, it } from "vitest"
import { ALL_USERS, PolicyUserAlice, PolicyUserBob, PolicyUserJohn, PolicyUserRoot } from "./policies.js"
import { type TestCtx, assertPermission, getTestCtx } from "./testSetup.js"
import { PolicyAct, PolicyEft, PolicyObj } from "./types.js"

let ctx: TestCtx
let ef: Enforcer
beforeAll(async () => {
	ctx = await getTestCtx()
	ef = ctx.ef
})

describe("policies", () => {
	for (const u of ALL_USERS) {
		it(`groups of user ${u}`, async () => {
			expect(await ef.getImplicitRolesForUser(u)).toMatchSnapshot()
		})
	}
})

describe("permissions", () => {
	// ---user PolicyUserRoot---
	// ==>denied
	assertPermission({
		sub: PolicyUserRoot,
		obj: PolicyObj.Movie,
		act: PolicyAct.Read,
	})(false)

	// ---user PolicyUserAlice---
	// ==>allowed
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Read,
	})(true)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Read,
		explicit: true,
	})(true)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Delete,
	})(true)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Delete,
		explicit: true,
	})(true)
	// ==>denied
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Create,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Create,
		explicit: true,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Update,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.Movie,
		act: PolicyAct.Update,
		explicit: true,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.MovieBrandedData,
		act: PolicyAct.Create,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.MovieBrandedData,
		act: PolicyAct.Create,
		explicit: true,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.MovieBrandedData,
		act: PolicyAct.Delete,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.MovieBrandedData,
		act: PolicyAct.Delete,
		eft: PolicyEft.Deny,
	})(false)
	assertPermission({
		sub: PolicyUserAlice,
		obj: PolicyObj.MovieLocalizedData,
		act: PolicyAct.Update,
	})(false)

	// ---user PolicyUserBob---
	// ==>denied
	assertPermission({
		sub: PolicyUserBob,
		obj: PolicyObj.Movie,
		act: PolicyAct.Read,
	})(false)
	assertPermission({
		sub: PolicyUserBob,
		obj: PolicyObj.MovieBrandedData,
		act: PolicyAct.Create,
	})(false)
	assertPermission({
		sub: PolicyUserBob,
		obj: PolicyObj.MovieLocalizedData,
		act: PolicyAct.Create,
	})(false)

	// ---user PolicyUserJohn---
	// ==>denied
	assertPermission({
		sub: PolicyUserJohn,
		obj: PolicyObj.Movie,
		act: PolicyAct.Read,
	})(false)
	assertPermission({
		sub: PolicyUserJohn,
		obj: PolicyObj.MovieBrandedData,
		act: PolicyAct.Create,
	})(false)
})
