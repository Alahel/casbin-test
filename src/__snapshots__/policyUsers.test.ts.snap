// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`policies > groups of user u:alice 1`] = `
[
  "g:redac_fr",
  "r:redac",
]
`;

exports[`policies > groups of user u:bob 1`] = `
[
  "g:b2b_ridgefield",
  "r:b2b_mavens",
]
`;

exports[`policies > groups of user u:john 1`] = `
[
  "g:b2b_appdot",
  "g:redac_fr",
  "r:redac",
]
`;

exports[`policies > groups of user u:root 1`] = `[]`;
